examples/model_compress/pruning/pruning_bert_glue.py

```diff
@@ -0,0 +1,355 @@
+    model.train()
+    if teacher_model:
+        teacher_model.eval()
+    current_step = 0
+    best_result = 0
+
+    for current_epoch in range(max_epochs if max_epochs else 1):
+        for batch in train_dataloader:
+            batch.to(device)
+            optimizer.zero_grad()
+            outputs = model(**batch)
+            loss = outputs.loss
+            if teacher_model:
+                distil_loss = F.kl_div(F.log_softmax(outputs.logits / 2, dim=-1), F.softmax(teacher_model(**batch).logits / 2, dim=-1), reduction='batchmean') * (2 ** 2)
+                loss = 0.1 * loss + 0.9 * distil_loss
+            loss = criterion(loss, None)
+            loss.backward()
+            optimizer.step()
+            if lr_scheduler:
+                lr_scheduler.step()
+            current_step += 1
+            if current_step % 1000 == 0 or current_step % len(train_dataloader) == 0:
+                result = evaluation(model)
+                with (log_path).open('a+') as f:
+                    f.write('[{}] Epoch {}, Step {}: {}\n'.format(time.asctime(time.localtime(time.time())), current_epoch, current_step, result))
+                if save_best_model and best_result < result[0]:
+                    assert save_path is not None
+                    torch.save(model.state_dict(), save_path)
+                    best_result = result[0]
+            if max_steps and current_step >= max_steps:
+                return
+
+
+def evaluation(model):
+    training = model.training
+    model.eval()
+    is_regression = task_name == 'stsb'
+    metric = load_metric('glue', task_name)
+
+    for batch in validate_dataloader1:
+        batch.to(device)
+        outputs = model(**batch)
+        predictions = outputs.logits.argmax(dim=-1) if not is_regression else outputs.logits.squeeze()
+        metric.add_batch(
+            predictions=predictions,
+            references=batch['labels'],
+        )
+    result = metric.compute()
+
+    if validate_dataloader2:
+        for batch in validate_dataloader2:
+            batch.to(device)
+            outputs = model(**batch)
+            predictions = outputs.logits.argmax(dim=-1) if not is_regression else outputs.logits.squeeze()
+            metric.add_batch(
+                predictions=predictions,
+                references=batch['labels'],
+            )
+        result = {'matched': result, 'mismatched': metric.compute()}
+
+        model.train(training)
+        return (result['matched']['accuracy'] + result['mismatched']['accuracy']) / 2, result
+
+    model.train(training)
+    return result.get('f1', result.get('accuracy', None)), result
+
+
+# using huggingface native loss
+def fake_criterion(outputs, targets):
+    return outputs
+
+
+if __name__ == '__main__':
+    ##########################################################################################################################
```

Comment:
How long does it take to run this script? Any chance we can write it as a tutorial?

If it takes long, maybe we can put a flag indicator upfront. By default, it runs in a quick development mode. When this flag is turned off, it runs in the full mode which runs the whole training.

Reply:
it might take 6~8 hours on A100, I can have a try to write it as a tutorial.